pinot-controller/src/test/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManagerTestNew.java

```diff
@@ -0,0 +1,1617 @@
+
+    tableConfig = makeTableConfig(tableName, nReplicas, KAFKA_OFFSET, DUMMY_HOST, DEFAULT_SERVER_TENANT,
+        DEFAULT_STREAM_ASSIGNMENT_STRATEGY);
+    nPartitions = 4;
+    segmentManager._partitionAssignmentGenerator.setConsumingInstances(instances);
+    PartitionAssignment expectedPartitionAssignment =
+        segmentManager._partitionAssignmentGenerator.generatePartitionAssignment(tableConfig, nPartitions);
+
+    // set up a happy path - segment is present in ideal state, is CONSUMING, metadata says IN_PROGRESS
+    idealState = clearAndSetupHappyPathIdealState(idealStateBuilder, segmentManager, tableConfig, nPartitions);
+
+    final int nInstancesChanged = 0;
+    final int nPartitionsIncreased = 1;
+    final int nInstancesChangedAndNPartitionsIncreased = 2;
+    final int[] numInstancesSet = new int[]{4, 6, 8, 10};
+
+    // randomly introduce an error condition and assert that we repair it
+    Random random = new Random(seed);
+    for (int run = 0; run < 200; run++) {
+
+      final boolean step3NotDone = random.nextBoolean();
+
+      final boolean tooSoonToCorrect = random.nextBoolean();
+      if (tooSoonToCorrect) {
+        segmentManager.tooSoonToCorrect = true;
+      }
+
+      if (step3NotDone) {
+        // failed after completing step 2: hence not present in ideal state
+
+        final boolean prevMetadataNull = random.nextBoolean();
+
+        if (prevMetadataNull) {
+          // very first seq number of this partition i.e. prev metadata null
+          clearAndSetupInitialSegments(idealStateBuilder, segmentManager, tableConfig, nPartitions);
+
+          // generate the error scenario - remove segment from ideal state, but metadata is IN_PROGRESS
+          int randomlySelectedPartition = random.nextInt(nPartitions);
+          Map<String, LLCSegmentName> partitionToLatestSegments =
+              segmentManager._partitionAssignmentGenerator.getPartitionToLatestSegments(idealState);
+          LLCSegmentName llcSegmentName = partitionToLatestSegments.get(String.valueOf(randomlySelectedPartition));
+          idealStateBuilder.removeSegment(llcSegmentName.getSegmentName());
+
+          // get old state
+          Assert.assertNull(idealState.getRecord().getMapFields().get(llcSegmentName.getSegmentName()));
+          nPartitions = expectedPartitionAssignment.getNumPartitions();
+          IdealState idealStateCopy = new IdealState(
+              (ZNRecord) znRecordSerializer.deserialize(znRecordSerializer.serialize(idealState.getRecord())));
+          Map<String, Map<String, String>> oldMapFields = idealStateCopy.getRecord().getMapFields();
+
+          if (tooSoonToCorrect) {
+            segmentManager.validateLLCSegments(tableConfig, idealState, nPartitions);
+            // validate nothing changed and try again with disabled
+            Assert.assertEquals(oldMapFields, idealState.getRecord().getMapFields());
+            segmentManager.tooSoonToCorrect = false;
```

Comment:
I set it back to false immediately, because I wanted it to try again with not tooSoonToCorrect, before going back up and introducing more error conditions. I did that because I wanted to avoid having to keep track of which partitions already have the error condition